### ISSUE TITLE
Wrong SOCKS5 protocol implemention. Socks5 with forward_socks not sending requests

### ISSUE THREAD

**Issue by juouy** [ronf/asyncssh#396]
Hi,

The socks forward proxy for SSH connection has wrong SOCKS5 protocol implementation and is returning wrong response.

Take a look at this simple code which connects to SSH server and opens local SOCKS5 proxy on our machine where we run script.

```
async def ssh_connect():
    conn, client = await asyncio.wait_for(asyncssh.create_connection(
        None, 'YOUR_IP_ADDRESS',
        port=22,
        known_hosts=None,
        username='YOUR_USERNAME',
        password='YOUR_PASSWORD'),
        100
    )

    async with conn:
        print("Connected...")
        listener = await asyncio.wait_for(conn.forward_socks('127.0.0.1', 1080), 600)
        return await listener.wait_closed()

try:
    asyncio.get_event_loop().run_until_complete(ssh_connect())
except (OSError, asyncssh.Error) as exc:
    print(str(exc))
```

This works fine with socks4 and socks4a using following command:

`curl google.com --socks4 127.0.0.1:1080`
`curl google.com --socks4a 127.0.0.1:1080`

However SOCKS5 does not work:

`curl google.com --socks5 127.0.0.1:1080`

The reason is quite simple. Your implementation is returning wrong answer to SOCKS CONNECT command:
https://github.com/ronf/asyncssh/blob/e1c36548fcd2de23890a59e5583c9f08d2b09f1e/asyncssh/socks.py#L42-L44

There's no way this is correct response. For more information you can look on RFC. 
The quick dirty fix to have it working is replacing the SOCKS5_OK_RESPONSE to:
```

SOCKS5_OK_RESPONSE      = bytes((SOCKS5, SOCKS5_OK, 0,
                                 SOCKS5_ADDR_HOSTNAME, 0, 0, 0, 0, 0, 0))
```

Please fix it at least the dirty way. 

In case you need it, here you have an example of SOCKS5 server in python with asyncio:
https://gist.github.com/tcalmant/770511b420c1139fe1cc88c2685f2903

**Comment by ronf** [comment]
Thank you for the report.

The current implementation in AsyncSSH is compliant with [RFC 1928](https://datatracker.ietf.org/doc/html/rfc1928)  and what you're suggesting to change it to would not be. Notice that the ATYP first in the AsyncSSH SOCK5 response is SOCKS5_ADDR_HOSTNAME, not SOCKS5_ADDR_IPV4 or SOCKS5_ADDR_IPV6. As such, that makes the byte following ATYP a length field, and since the length returned is 0 that means there are no more BND.ADDR bytes after that. The last two 0 bytes are the BND.PORT value. In other words, the response here is returning an empty hostname for BND.ADDR, which is perfectly legal according to the RFC.

I've tested the AsyncSSH SOCKS5 support against Firefox, and it works fine as-is. I wasn't aware of the problem you mentioned with curl, but unless I'm missing something this is a problem in curl. Looking at their implementation at https://github.com/curl/curl/blob/fd84db600d2ace530ad17616cb21d8b78b02840d/lib/socks.c it seems like they are trying to handle a hostname, properly reading the length byte and calculating the total response length based on that. So, I'm not quite sure what's going wrong there that causes their client to hang. I think it may have to do with the fact that curl expects the minimum response to be 10 bytes, and a response with an empty hostname is only 7 bytes. It really should only be reading 7 bytes before trying to compute the actual length of the response. I did a test where I returned a 2-byte hostname instead of an empty one and it still failed, but as soon as I returned 3 bytes  or more, it succeeded, properly removing the correct number of bytes for the variable-length hostname response before continuing.

Changing the "address" field to be four 0 bytes instead of one would only make sense if you also changed the ATYP field from SOCKS5_ADDR_HOSTNAME to SOCKS5_ADDR_IPV4. It looks like that makes curl happy, but it's kind of "dirty" as you said to return an all-zero IPv4 address even when possibly connecting via IPv6. Returning an empty hostname was more neutral in that regard, and also happened to be a little bit shorter.

One of the things that makes it difficult to implement what SOCKS intended here is that the SSH port forwarding mechanism doesn't return the local addr and port of the remote connection. So, even if I wanted to do this "right" and return what the outbound socket's local addr & port were, the SSH protocol doesn't send that information back to the SSH client.

As you noted, you can work around this problem in curl by requesting SOCKSv4 or SOCKSv4a. Since there's no support for authentication in the SSH SOCKS proxies (in either AsyncSSH or OpenSSH), you're really not losing anything by choosing one of those.

**Comment by juouy** [comment]
Your implementation does not work with other proxy related tools either such as ProxyCap too
Open wireshark and see how regular ssh is responding. https://i.imgur.com/cXQRlM5.png


Section 6 of RFC describes replies to different request types:

```
CONNECT

   In the reply to a CONNECT, BND.PORT contains the port number that the
   server assigned to connect to the target host, while BND.ADDR
   contains the associated IP address.  The supplied BND.ADDR is often
   different from the IP address that the client uses to reach the SOCKS
   server, since such servers are often multi-homed.  It is expected
   that the SOCKS server will use DST.ADDR and DST.PORT, and the
   client-side source address and port in evaluating the CONNECT
   request.
```

Let's quote some important lines.

```
while BND.ADDR
   contains the associated IP address.
``` 

If RFC states that it should be IP address. Why is the address type in your response then set to SOCKS5_ADDR_HOSTNAME?

Yes in section 5 of RFC it states that it can be domain too, however in replies section 6 for CONNECT case it clearly states that it should be IPV4 in this particular case .


The only reason my quick dirty fix works is because it emulates IPV4 response. I forgot to set to atype to IPV4. I have added 4 bytes after ATYPE then 2 bytes for port. 

It most universal implementation to expect 4 bytes of IP, as RFC clearly states that ATYPE should be IP address.

Here you have response from bitvise ssh tunneler and openssh:
https://i.imgur.com/7tCmsfx.png
https://i.imgur.com/cXQRlM5.png

They both set atype to IPV4 and then bitvise fills it with real address, while openssh fills it with zeros. But still there are 4 empty bytes.

> So, I'm not quite sure what's going wrong there that causes their client to hang. I think it may have to do with the fact that curl expects the minimum response to be 10 bytes, and a response with an empty hostname is only 7 bytes. It really should only be reading 7 bytes

Nothing appearntly is going wrong with curl.   
Here you have someone reporting this as a bug:
https://github.com/curl/curl/issues/3165


> In other words, the response here is returning an empty hostname for BND.ADDR, which is perfectly legal according to the RFC.

Show me that part..  BND.ADDR should be set  of IPV4 or IPV6 and it should be filled with SOMETHING.  Even if it's blank data.

**Comment by ronf** [comment]
Given multiple tools are having a problem with this, I'll consider making a change here. However, unlike a "real" SOCKS server which always _knows_ the source address and port (and address type) it used to make the requested outbound connection, an SSH client implementing SOCKS like this does NOT know the source address and port used to connect upstream, and it CAN'T know it due to the format of the SSH messages used when opening a "direct-tcpip" channel. The SSH server which makes this outbound connection simply does not return that information to the client. In fact, a client can't even know the proper ATYP to fill in here in cases where the ATYP from the SOCKS client is of type hostname, as the DNS resolution of that hostname doesn't happen on the SSH client. It also happens only on the server, and the server doesn't return what type of connection it ended up making. If Bitvise is filling something in here, I'm guessing it is probably filling in the source IP and port of the upstream connection it makes to the SSH server, and not the IP and port the server used to make the upstream connection the SOCKS client is requesting.

Regarding curl, sending the extra 0x00 bytes you suggested but leaving the ATYP set to hostname would _not_ have worked properly. It would have avoided the curl client hanging, but the three extra 0x00 bytes would have ended up being treated as the first three bytes of the data tunneled via SOCKS, which would likely break applications trying to use this. As you noted elsewhere in your response, you need to set the ATYP to IPv4 if you want to send a fixed set of 4 0x00 bytes for the address. However, I wonder if that could cause problems for some clients if they made a SOCKS CONNECT request explicitly asking to connect to an IPv6 address. This should also "never happen" on a "real" SOCKS server. It seems to me to really be on the safe side one would need to keep the ATYP the same as what the client request contained when the client sets ATYP to IPv4 or IPv6. For cases where the client sets the ATYP to hostname, it might be ok to always pretend the hostname resolved to an IPv4 address return an ATYP of IPv4, but in the case where the client asks specifically to make an IPv6 connection, I could see forcing an ATYP of IPv4 in the response could lead to problems, just as you're claiming returning hostname does for some clients.

You say Bitvise and OpenSSH always return an ATYP of IPv4, but did you test sending a SOCKS CONNECT request with an IPv6 target, or with a hostname like ipv6.google.com that only resolves to an IPv6 address?

**Comment by ronf** [comment]
Here's a potential fix which preserves the ATYP in the response for IPv4 and IPv6 SOCKS CONNECT requests, and sends an IPv4 response in the case where the request ATYP is a hostname. As before, it fills in all zeroes for BND.ADDR and BND.PORT, but the length of the address is set to match the chosen response ATYP (which will now always be either IPv4 or IPv6):

```patch
diff --git a/asyncssh/socks.py b/asyncssh/socks.py
index 8d2f61e..9858eb4 100644
--- a/asyncssh/socks.py
+++ b/asyncssh/socks.py
@@ -39,8 +39,7 @@ SOCKS5_ADDR_HOSTNAME    = 0x03
 SOCKS5_ADDR_IPV6        = 0x04

 SOCKS4_OK_RESPONSE      = bytes((0, SOCKS4_OK, 0, 0, 0, 0, 0, 0))
-SOCKS5_OK_RESPONSE      = bytes((SOCKS5, SOCKS5_OK, 0,
-                                 SOCKS5_ADDR_HOSTNAME, 0, 0, 0))
+SOCKS5_OK_RESPONSE_HDR  = bytes((SOCKS5, SOCKS5_OK, 0))

 _socks5_addr_len = { SOCKS5_ADDR_IPV4: 4, SOCKS5_ADDR_IPV6: 16 }

@@ -54,6 +53,7 @@ class SSHSOCKSForwarder(SSHLocalForwarder):
         self._inpbuf = b''
         self._bytes_needed = 2
         self._recv_handler = self._recv_version
+        self._addrtype = 0
         self._host = ''
         self._port = 0

@@ -73,7 +73,11 @@ class SSHSOCKSForwarder(SSHLocalForwarder):
     def _send_socks5_ok(self):
         """Send SOCKS5 success response"""

-        self._transport.write(SOCKS5_OK_RESPONSE)
+        addrlen = _socks5_addr_len[self._addrtype] + 2
+
+        self._transport.write(SOCKS5_OK_RESPONSE_HDR +
+                              bytes((self._addrtype,)) +
+                              addrlen * b'\0')

     def _recv_version(self, data):
         """Parse SOCKS version"""
@@ -144,12 +148,14 @@ class SSHSOCKSForwarder(SSHLocalForwarder):
             if data[3] == SOCKS5_ADDR_HOSTNAME:
                 self._bytes_needed = 1
                 self._recv_handler = self._recv_socks5_hostlen
+                self._addrtype = SOCKS5_ADDR_IPV4
             else:
                 addrlen = _socks5_addr_len.get(data[3])

                 if addrlen:
                     self._bytes_needed = addrlen
                     self._recv_handler = self._recv_socks5_addr
+                    self._addrtype = data[3]
                 else:
                     self.close()
         else:
```

**Comment by juouy** [comment]
> You say Bitvise and OpenSSH always return an ATYP of IPv4, but did you test sending a SOCKS CONNECT request with an IPv6 target, or with a hostname like ipv6.google.com that only resolves to an IPv6 address?

I dont have access to IPv6 setup right now.
Take a look here

https://github.com/openssh/openssh-portable/blob/464ba22f1e38d25402e5ec79a9b8d34a32df5a3f/channels.c#L1505-L1515

**Comment by juouy** [comment]
I managed to get access to IPv6. Here is how bitvise behaves:

Connect request to ipv6.google.com and connect response

https://i.imgur.com/tSTDQMu.png

**Comment by ronf** [comment]
Thanks for the additional information.

I did some further testing of the Bitvise client, and it appears that it can return either an IPv4 or IPv6 ATYP in its response, but it depends on whether its connection to the SSH server is IPv6 or not, and is not tied to the ATYP in the SOCKS CONNECT request. The IP address it fills in is actually the server IP address of the outbound SSH connection, and from what I can tell the port number reported back is a made-up (I'm guessing randomly chosen) value that doesn't seem to relate to either the SSH connection's ports or the SOCKS connection's ports.

I'll need to do some additional testing of OpenSSH to determine if it really hardcodes the IPv4 ATYP in its response even when connecting to the SSH server via IPv6, but given that it is not filling in any server address information, I suspect it might be doing that. Even though it looks like both tools can send back an IPv4 ATYP even when the client is asking for an IPv6 connection, I still worry about some SOCKS clients not liking that.

**Comment by ronf** [comment]
Ok - I've provided a fix for this in commit 3d10f9f in the "develop" branch. It preserves the request ATYP in the response whenever the request sends an IPv4 or IPv6 address, and responds with an ATYP of IPv4 for requests with a hostname, to maximize compatibility with older clients that might not handle IPv6 addresses. In all cases, the response address and port will be all zeroes, since there's no way for the SSH client to know the actual bound address and port assigned by the SSH server.

Thanks for the information you provided about the compatibility issues you saw with various SOCKS clients!

**Comment by juouy** [comment]
Thank you

**Comment by ronf** [comment]
This change is now available in AsyncSSH version 2.7.1.
